### PR TITLE
Include annotations on test methods in output

### LINKED ISSUE #1: Include annotations along with method names to allow filtering
Since instrumentation allows filtering by annotation, it would be useful to include list of all annotations for each test method and the whole test class.

Also, in what format should the data be stored? CSV with different separators for columns and lists, or maybe JSON?

### PR DESCRIPTION
**Pull request by drewhannay**
DexParser now has a new top-level entry point method
that will return not only the fully qualified test
names found in the given apk, but also the fully qualified
names of any annotations that apply to the test (either
class or method level annotations for that test).

Note that the parser api that writes output to a file
on disk has not currently been changed; it still only
writes the fully qualified test names and ignores the
annotations. Annotation retrieval is only supported
when using dex-test-parser as a Java dependency, not
when used as a command line tool. It should be easy for
someone to write a wrapper around dex-test-parser to output
the annotations via a command line tool if they need it,
and this saves dex-test-parser from needing to maintain
a more complicated serialized format of the annotation
output.

This commit has the side effect of updating to Kotlin
1.1.2 and fixing all the new compiler warnings.

Fixes #1

### CODE REVIEW

File: parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * For example,
+ * "Lorg/junit/Test;"
+ * would become:
+ * "org.junit.Test"
+ */
+fun formatDescriptor(descriptor: String): String {
+    return descriptor
+            // strip off the "L" prefix
+            .substring(1)
```
Comment (kkoser) [inline parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt:21]: We should probably either add more comments about the expected format of the input, or do some checks to see the format of the string here. If I pass in "org.junit.Test" to this method, I'm going to get back "rg.junit.Tes" which could cause confusion

File: parser/src/main/kotlin/com/linkedin/dex/parser/AnnotationUtils.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+fun DexFile.getAnnotationsDirectory(classDefItem: ClassDefItem): AnnotationsDirectoryItem? {
+    if (hasAnnotations(classDefItem)) {
+        return AnnotationsDirectoryItem.create(byteBuffer, classDefItem.annotationsOff)
+    } else {
+        return null
+    }
+}
+
+/**
+ * @return A list of descriptors containing all class-level annotations in the given [AnnotationsDirectoryItem]
```
Comment (kkoser) [inline parser/src/main/kotlin/com/linkedin/dex/parser/AnnotationUtils.kt:36]: Can we be consistent with referring to items either by name or by type? I see both here

File: parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt [hunk 3]
```diff
@@ -5,67 +5,30 @@
...
-    // strip off the ";"
-    className = className.dropLast(1)
-    // the instrument command expects test class names and method names to be separated by a "#"
-    className += "#"
-
-    return className
+/**
+ * Find methodIds we care about: any method in the class which is annotated
+ */
+private fun DexFile.findMethodIds(): (ClassDefItem, AnnotationsDirectoryItem?) -> List<MethodIdItem> {
```
Comment (kkoser) [inline parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt:28]: why does this need to be a method? cant this just be a private variable? Theres no need to create the lambda on the fly

File: parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            // strip off the ";"
+            .dropLast(1)
+}
+
+/**
+ * Extract the fully qualified class name from a [ClassDefItem] and format it for use with Android instrumentation tests
+ *
+ * @see [formatDescriptor]
+ */
+fun DexFile.formatClassName(classDefItem: ClassDefItem): String {
```
Comment (kkoser) [inline parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt:33]: Not sure the naming here makes sense to me... it doesn't so much format it for use with instrumentation tests as format it to be combined with a test name for instrumentation tests. What do you think about naming it more specifically? Not a huge deal

**Review by kkoser — APPROVED** [review]
LGTM

**Review by jasmsu — APPROVED** [review]

File: parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt (hunk 1 above)
Comment (drewhannay) [reply parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt:21]: "Class descriptor" is a specific term which means this format. I tried to use "descriptor" vs "name" consistently throughout the codebase so i think this one is okay.

File: parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt (hunk 4 above)
Comment (drewhannay) [reply parser/src/main/kotlin/com/linkedin/dex/parser/FormatUtils.kt:33]: Yeah, I don't love this name either, but I also don't have any better suggestions

File: parser/src/main/kotlin/com/linkedin/dex/parser/AnnotationUtils.kt (hunk 2 above)
Comment (drewhannay) [reply parser/src/main/kotlin/com/linkedin/dex/parser/AnnotationUtils.kt:36]: All of them are referring to type, which I think is most readable, EXCEPT for the one where the type is `List<String>`. I think changing them all to name makes four of the five less readable, and changing the last one to type makes that one completely UN-readable. So I think I prefer to leave it as-is

File: parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt (hunk 3 above)
Comment (drewhannay) [reply parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt:28]: I like that idea, and just tried to make the change, but currently this is an extension method on `DexFile` and needs access to the `methodIds` property on `DexFile`, so i think it needs to stay a method

**Pull request merged**